packages/commerce-sdk-react/src/hooks/ShopperBaskets/cache.ts

```diff
@@ -146,39 +72,287 @@ const TODO = (method: keyof Client) => {
+                                      response,
+                                      oldData
+                                  )
+                          }
+                      ]
+                    : [])
+            ]
+        }
+    },
     createShipmentForBasket: TODO('createShipmentForBasket'),
-    deleteBasket: (customerId, {parameters}) => ({
-        // TODO: Convert invalidate to an update that removes the matching basket
-        ...invalidateCustomerBasketsQuery(customerId, parameters),
-        remove: [
-            and(
-                matchParameters(toGetBasketParameters(parameters)),
-                matchesPath(getBasketPath(parameters))
-            )
-        ]
-    }),
-    mergeBasket: updateBasketWithResponseBasketId,
-    removeCouponFromBasket: updateBasket,
+    deleteBasket(customerId, {parameters}) {
+        return {
+            // TODO: Convert invalidate to an update that removes the matching basket
+            ...invalidateCustomerBasketsQuery(customerId, parameters),
+            remove: [
+                {queryKey: getBasket.queryKey(parameters)},
+                ...(customerId
+                    ? [{queryKey: getCustomerBaskets.queryKey({...parameters, customerId})}]
+                    : [])
+            ]
+        }
+    },
+    mergeBasket(customerId, {parameters}, response) {
+        const {basketId} = response
+        return {
+            update: [
+                {queryKey: getBasket.queryKey(parameters)},
+                ...(customerId && basketId
+                    ? [
+                          {
+                              // @ts-ignore
+                              queryKey: getCustomerBaskets.queryKey({
+                                  ...parameters,
+                                  basketId,
+                                  customerId
+                              }),
+                              updater: (oldData: CustomerBasketsResult | undefined) =>
+                                  customerBasketsUpdater(
+                                      {...parameters, basketId},
+                                      response,
+                                      oldData
+                                  )
+                          }
+                      ]
+                    : [])
+            ]
+        }
+    },
+    removeCouponFromBasket(customerId, {parameters}, response) {
+        return {
+            update: [
+                {queryKey: getBasket.queryKey(parameters)},
+                ...(customerId
+                    ? [
+                          {
+                              queryKey: getCustomerBaskets.queryKey({...parameters, customerId}),
+                              updater: (oldData: CustomerBasketsResult | undefined) =>
+                                  customerBasketsUpdater(parameters, response, oldData)
+                          }
+                      ]
+                    : [])
```

Comment:
I see a pattern where some updates are based on conditions. If I read this code correctly, this is "if customerId is defined, do the update..." The pattern here is somewhat difficult to read, can we introduce a `predicate` property to make this a bit more readable?

```suggestion
                {
                        predicate: () => !!customerId
                        queryKey: getCustomerBaskets.queryKey({...parameters, customerId}),
                        updater: (oldData: CustomerBasketsResult | undefined) =>
                            customerBasketsUpdater(parameters, response, oldData)
                 }

```

Reply:
React query only has `predicate` for invalidate/remove. For `update`, there's just `queryKey` and `updater?`. We could introduce our own `predicate` logic (but probably not in this PR) and it would likely break TypeScript (the `updater` scope wouldn't know what the `predicate` is asserting, and wouldn't narrow types to eliminate `undefined`).

Reply:
ok, got it. Let's not do it in this PR.